x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authz/permission/FieldPermissionsTests.java

```diff
@@ -82,23 +85,63 @@ public void testFieldPermissionsIntersection() {
         }
     }
 
+    public void testMultipleLimiting() {
+        // Basic test for a number of permission definitions
+        FieldPermissions fieldPermissions = FieldPermissions.DEFAULT;
+        final int nSets = randomIntBetween(2, 8);
+        final FieldPermissionsDefinition fieldPermissionsDefinition = fieldPermissionDef(
+            new String[] { "f1", "f2", "f3*" },
+            new String[] { "f3" }
+        );
+        for (int i = 0; i < nSets; i++) {
+            fieldPermissions = fieldPermissions.limitFieldPermissions(new FieldPermissions(fieldPermissionsDefinition));
+        }
+        final List<FieldPermissionsDefinition> fieldPermissionsDefinitions = fieldPermissions.getFieldPermissionsDefinitions();
+        assertNonNullFieldPermissionDefinitions(fieldPermissionsDefinitions, nSets);
+        fieldPermissionsDefinitions.forEach(fpd -> assertThat(fpd, equalTo(fieldPermissionsDefinition)));
+        assertThat(fieldPermissions.grantsAccessTo(randomFrom("f1", "f2", "f31")), is(true));
+        assertThat(fieldPermissions.grantsAccessTo(randomFrom("f1", "f2", "f3")), is(false));
+
+        // More realistic intersection
+        fieldPermissions = FieldPermissions.DEFAULT;
+        fieldPermissions = fieldPermissions.limitFieldPermissions(
+            new FieldPermissions(fieldPermissionDef(new String[] { "f1", "f2", "f3*", "f4*" }, new String[] { "f3" }))
+        );
+        fieldPermissions = fieldPermissions.limitFieldPermissions(
+            new FieldPermissions(fieldPermissionDef(new String[] { "f2", "f3*", "f4*", "f5*" }, new String[] { "f4" }))
+        );
+        fieldPermissions = fieldPermissions.limitFieldPermissions(
+            new FieldPermissions(fieldPermissionDef(new String[] { "f3*", "f4*", "f5*", "f6" }, new String[] { "f5" }))
+        );
+        assertNonNullFieldPermissionDefinitions(fieldPermissions.getFieldPermissionsDefinitions(), 3);
+
+        assertThat(fieldPermissions.grantsAccessTo(randomFrom("f1", "f2", "f5", "f6") + randomAlphaOfLengthBetween(0, 10)), is(false));
+        assertThat(fieldPermissions.grantsAccessTo("f3"), is(false));
+        assertThat(fieldPermissions.grantsAccessTo("f4"), is(false));
+
+        assertThat(fieldPermissions.grantsAccessTo("f3" + randomAlphaOfLengthBetween(1, 10)), is(true));
+        assertThat(fieldPermissions.grantsAccessTo("f4" + randomAlphaOfLengthBetween(1, 10)), is(true));
+    }
```

Comment:
Can you help me understand why the following passes
```
        fieldPermissions = FieldPermissions.DEFAULT;
        fieldPermissions = fieldPermissions.limitFieldPermissions(
            new FieldPermissions(fieldPermissionDef(new String[] { "f1", "f2" }, new String[] { "" }))
        );

        assertThat(fieldPermissions.grantsAccessTo("f1"), is(true));
        assertThat(fieldPermissions.grantsAccessTo("f2"), is(true));
```
but the following does not
```
        fieldPermissions = FieldPermissions.DEFAULT;
        fieldPermissions = fieldPermissions.limitFieldPermissions(
            new FieldPermissions(fieldPermissionDef(new String[] { "f1" }, new String[] { "" }))
        );
        fieldPermissions = fieldPermissions.limitFieldPermissions(
            new FieldPermissions(fieldPermissionDef(new String[] { "f2" }, new String[] { "" }))
        );
        assertThat(fieldPermissions.grantsAccessTo("f1"), is(true));
        assertThat(fieldPermissions.grantsAccessTo("f2"), is(true));
```
Shouldn't those two cases behave identical since we are saying grant access to "f1" and grant access to "f2" ?

Reply:
The 1st test case is equivalent to a SimpleRole (e.g. for realm users), in role descriptor form it is like
```
{
  "field_security": {
    "grant": [ "f1", "f2" ],
    "except": [ "" ]
  }
}
```
It is pretty clear that the user is granted for access both `f1` and `f2`.

---
The 2nd test case is equivalent to a LimitedRole (for API keys), in role descriptor form it is like
```
{
  "key_role": {
    "field_security": {
      "grant": [ "f1" ],
      "except": [ "" ]
    }
  },
  "limited_by": {
    "field_security": {
      "grant": [ "f2" ],
      "except": [ "" ]
    }
  }
}
```
Even though the key's role descriptor says it can access filed `f1`, it is *not* allowed by the limiting user's role (only allow `f2`). Similarly, `f2` is *not* accessible because the key's role does *not* allow it. 
For any field to be accessible, it must be allowed by both key role and the limiting user role.

Reply:
Thanks, presented that way, makes perfect sense !